src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -2680,21 +2691,68 @@ class ShenandoahUpdateHeapRefsTask : public AbstractGangTask {
     if (_concurrent) {
       ShenandoahConcurrentWorkerSession worker_session(worker_id);
       ShenandoahSuspendibleThreadSetJoiner stsj(ShenandoahSuspendibleWorkers);
-      do_work();
+      do_work(worker_id);
     } else {
       ShenandoahParallelWorkerSession worker_session(worker_id);
-      do_work();
+      do_work(worker_id);
     }
   }
 
 private:
-  void do_work() {
+  void do_work(uint worker_id) {
     ShenandoahHeapRegion* r = _regions->next();
     ShenandoahMarkingContext* const ctx = _heap->complete_marking_context();
+
     while (r != NULL) {
       HeapWord* update_watermark = r->get_update_watermark();
       assert (update_watermark >= r->bottom(), "sanity");
-      if (r->is_active() && !r->is_cset()) {
+
+      // Eventually, scanning of old-gen memory regions for the purpose of updating references can happen
+      // concurrently.  This can be done during concurrent evacuation of roots for example.
+      if (r->is_active() && (r->affiliation() == ShenandoahRegionAffiliation::OLD_GENERATION) && !r->is_cset()) {
+
+        // Note that we use this code even if we are doing an old-gen collection and we have a bitmap to
+        // represent marked objects within the heap region.
+        //
+        // It is necessary to process all objects rather than just the marked objects during update-refs of
+        // an old-gen region as part of an old-gen collection.  Otherwise, a subseqent update-refs scan of
+        // the same region will see stale pointers and crash.
+        //
```

Comment:
This is probably a bad idea. Unreachable objects are not guaranteed to have their Klass* intact, and may crash when accessed. Why would it crash during subsequent scans?